### ISSUE TITLE
So how does this work?

### ISSUE THREAD

**Issue by zilti** [AppImage/AppImageUpdate#95]
I'm trying to create an auto-updating appimage for an internally used application. But... there's no documentation. Zero. Nothing. Except a "TODO: Add tutorial". Is there any kind of example? What do I need to add to the AppImage, an XML file? .ini file? Where do I place it, how do I generate a zsync file?

Seriously, you need to get that sorted out *now*, or everyone will already have switched to flatpak, and switched for good.

**Comment by probonopd** [comment]
Thanks for your feedback.

* _I want my application to be updateable via external tools like AppImageUpdate_ --> All you need to do is to create a zsync file and embed the needed update information into your AppImage using `appimagetool -u`
* _I am using Travis CI on GitHub and want my application to be updateable via external tools like AppImageUpdate_ --> If you are using `linuxdeployqt` it will be fully automatic - nothing extra to do
* _I want my GUI application to be self-updateable from within the application. My application uses Qt_ --> You should soon be able to include a Qt component into your application, @TheAssassin has the details
* _I want my GUI application to be self-updateable from within the application. My application uses some other toolkit_ --> You can include `libappimageupdate` into your application, @TheAssassin has the details
* _I want my GUI application to be self-updateable from within the application but I don't want to touch my application's source code_ --> You can have a tiny wrapper plus AppImageUpdate bundled in your AppImage, @TheAssassin has the details

Which one describes your situation best?

**Comment by TheAssassin** [comment]
First of all, your question isn't worded very precisely. Please explain your idea of "auto updating". From our perspective, your primary use case of our libraries is "my AppImage shall be able to update itself". Whether this is automatic or not is out of scope of an update library. Therefore we try to avoid the term "auto update" and use "self-update" etc., as "auto update" could as well be used to describe an app store that wants to automatically fetch updates for external AppImage.

I don't understand your criticism regarding the "TODO". There's no "add tutorial". In fact, it states that the texts from the old readme should be copied. There is actually a link to the old README, which contains all the information: https://github.com/AppImage/AppImageUpdate/blob/master/README.md#update-information-overview. You can't simply claim there was no helpful text. I guess you're able to click a link...
Sure, this is a bit out of date, but if you had used Google, you could've found e.g., https://github.com/AppImage/AppImageSpec/blob/master/draft.md#update-information, which contains the up to date description of all the update information types.

Regarding the creation of the `.zsync` file, you're right, this doesn't necessarily get clear from the README, and the old README is outdated. In fact, appimagetool does it all for you. Check out its [usage](https://github.com/AppImage/AppImageKit#appimagetool-usage). As soon as you provide update information via `-u`, a suitable `.zsync` file is generated for you automatically. There's no dependencies on external tools.

Regarding how to use libappimageupdate, it depends on what you're trying to do.
Do you want to take over full control of the update process, e.g. to provide your own GUI, or have a CLI AppImage? Then you need to use libappimageupdate directly.
Do you have a UI app, ideally written in Qt? You can use our reusable update GUI which provides a widget that you simply need to `show()`, everything else is then automated. This is showcased in https://github.com/TheAssassin/AppImageLauncher/blob/master/src/update_main.cpp. It could probably use some love API wise, though. Feel free to create new issues with feature proposals.
You can also just bundle our official AppImages and call them as subprocesses (passing `$APPIMAGE` as the path).

So, please get a bit more precise in what you want/need, otherwise we can't help you either.

Oh, and please remember this is a free software project with very limited resources. We don't have a bunch of highly skilled developers working full time on these tools. Also, this is still beta software.

And rest assured, our own market analyses don't show any trend that people will leave for "alternatives". AppImage is unique in what it provides, Flatpak does so many things differently, and our users, both developers and users, prefer it therefore.

Next time, I'd recommend you to simply get in touch with us on IRC (#appimage on Freenode). This allows us to perform some requirements engineering (i.e., find out what you actually want to accomplish), and then guide you to a solution.

By the way, if you feel there's a lack of documentation: PRs welcome! I can explain you anything you want to know here as well, and it'd be highly appreciated if you could transform it into a wiki page or README or whatever format you think makes sense.

**Comment by probonopd** [comment]
I can see that for a newcomer the whole update system can be overwhelming, and I'd like us to create easy step-by-step instructions on what to do.

|GUI app?|Qt app?|Travis CI/GitLab?|Self-updating?|Can modify source?|Instructions|Howto/Example|
|-|-|-|-|-|-|-|
|Yes|Yes|Yes|No|*|Use `linuxdeployqt`, it will automatically embed update information and produce a zsync file||
|Yes|*|No|No|*|Use `appimagetool -u`, it will automatically embed update information and produce a zsync file||
|Yes|Yes|Yes|Yes|Yes|Include Qt component for updating the running AppImage||
|Yes|*|Yes|Yes|No|Include wrapper and bundle AppImageUpdate for updating the running AppImage||
|*|*|Yes|Yes|Yes|Include `libappimageupdate` for updating the running AppImage||
||||||||
||||||||
||||||||

* = Does not matter

**Comment by probonopd** [comment]
@TheAssassin can you fill us in on the __example needed__`? Then we could publish this somewhere, and maybe add a nice writeup/walktrough around it.

**Comment by zilti** [comment]
Ok, that info helped *a lot*, thank you! Also, yes I've seen that other link, but I didn't read it because I assumed it's obsolete (seeing how the link was labeled).

Right now, I'm at this:

```bash
appimagetool videocapture.AppDir videocapture.AppImage
echo "zsync|https://lyrion.ch/opensource/repositories/videocapture/uv/videocapture.AppImage.zsync" | dd of="videocapture.AppImage" bs=1 seek=33651 count=512 conv=notrunc 2>/dev/null
zsyncmake videocapture.AppImage
chmod +x videocapture.AppImage
```

Now as I understand it, the final step could be (since I want it to be self-updating) to add zsync-update to my AppImage, and call it with the url to the zsync file. So will this replace the file in-place? And does this directly affect the mounted AppImage, meaning I can proceed executing the main program?

**Comment by TheAssassin** [comment]
> appimagetool videocapture.AppDir videocapture.AppImage
> echo "zsync|https://lyrion.ch/opensource/repositories/videocapture/uv/videocapture.AppImage.zsync" | > dd of="videocapture.AppImage" bs=1 seek=33651 count=512 conv=notrunc 2>/dev/null
> zsyncmake videocapture.AppImage
> chmod +x videocapture.AppImage

You can do it like `appimagetool <...> -u "zsync|..."` in a single call. It will take care of everything.

**Comment by probonopd** [comment]
> Right now, I'm at this

Why not just this?

```
appimagetool videocapture.AppDir/usr/share/applications/*.desktop -u "zsync|https://lyrion.ch/opensource/repositories/videocapture/uv/videocapture.AppImage.zsync"
```

It should have the exact same effect.

The result is that your AppImage should be updateable using AppImageUpdate then. Now, if you want to have it _self-_ updateable you could

* Bundle AppImageUpdate inside your AppImage and have your application call it (it will give you an "alien-looking" Qt-based UI, or
* Build your own UI by consuming `libappimageupdate` inside your application

Which route would you like to go?

**Comment by zilti** [comment]
Ah, yes. I separated it onto multiple lines to comment out, because I had a problem at some point and wanted to rule out it's something with the appimage tools.

Bundling AppImageUpdate is fine; for this project neither is a super-fancy-looking updater required, nor is it alien to running on the system which is in every case so far a system with Plasma 5.

The only part of the puzzle missing at the moment is: what happens after AppImageUpdate has downloaded an update? Will the mounted image have changed as well already, or do I have to re-execute the AppImage? In case of the latter, what's the recommended way to do so, and what's the exit code? And (since AppImageUpdate ignores "-h"): can I somehow tell it to continue after updating instead of depending on the user clicking to launch?

**Comment by TheAssassin** [comment]
@probonopd meant "alien looking" as in "probably doesn't look like it's part of your main UI".

AppImageUpdate will simply create a file next to your current one with the new filename. If your new AppImage's filename is the same as the old one, the old AppImage will be renamed to `<name>.AppImage.zs-old`. That way, a running AppImage can continue to run.

If you need to get the exact filename of the new AppImage, you should consider using libappimageupdate, which provides that information reliably. If you need some fine grained control about AppImageUpdate's UI (e.g., provide your own option to restart the updated AppImage, deactivate the launch button or change its behavior, etc.), you won't get around using the library. AppImageUpdate is simply a standalone application that can work to some extent when bundled, but it's really limited to standalone use.

If you need update checks, appimageupdatetool (the CLI tool) has a `-j` flag. I'm not sure whether AppImageUpdate (the UI) implements that already, but as you say it ignores `-h`, it probably doesn't. Please open an issue over at AppImageUpdate's repo for features you are missing, and we can talk about implementing them.

**Comment by probonopd** [comment]
Your application should invoke `AppImageUpdate "${APPDIR}"` e.g., when the user selects "Check for updates..." from the Help menu. If AppImageUpdate is bundled inside the AppImage in `usr/bin`, then that instance is on the `$PATH` and will be launched.

"Updating" means "getting the binary delta and saving it in a _new_ AppImage, without changing or deleting the current (old) one. This allows for the user to have both versions and delete the old one only after having verified that the new one works.

**Comment by probonopd** [comment]
@TheAssassin I imagine what @zilti means is:

AppImageUpdate has a button to launch the new AppImage. This is useful e.g., if the user runs AppImageUpdate on another AppImage. But if AppImageUpdate is invoked from _within_ the (old) AppImage, then the user most likely does not want to launch the new AppImage _in addition_ to the (still running) old AppImage.

Hence, I would say, if AppImageUpdate was launched from inside an AppImage, then we should not have the "Launch" button but instead tell the user to quit the application and launch the new AppImage. (Do we have the info needed to do this?)

**Comment by TheAssassin** [comment]
I don't see why AppImageUpdate should focus on being included in other AppImages. It's just a bulk of libraries that the main application may even use itself. Bundling it is more of a "yeah, you can do it, if you're lazy" than an "officially supported use case".

**Comment by probonopd** [comment]
In case the application author doesn't want to deal with writing a GUI for it perhaps?

**Comment by TheAssassin** [comment]
I don't think it's good style to do this sort of stuff in a subprocess. But maybe you're right. Please write up what you would need to better support this use case. Then, we can think about how to do this. Perhaps we can introduce some environment variable the parent app has to set to make AppImageUpdate follow the other workflow.

**Comment by TheAssassin** [comment]
Just an update for you, @zilti, we started to work on docs for the entire AppImage project. We started with some user docs at https://docs.appimage.org/. API documentation for AppImageKit and AppImageUpdate is on the TODO list.

**Comment by technic** [comment]
One more thing to mention regarding *linuxdeployqt*. Wheny you run
`./linuxdeployqt <app> -appimage`
it internally executes `./appimagetool -g ...` which means it would try to guess the update zsync url. If you need more fine tuning in your AppImage, then you can call *appimagetool* tool directly, because it has more flags available. So you first call
`./linuxdeployqt -bundle-non-qt-libs ...`
to gather Qt staff to the AppDir, and then:
`./appimagetool -g ...`
to create your AppImage.

**Comment by TheAssassin** [comment]
... or you use [linuxdeploy](https://github.com/linuxdeploy/linuxdeploy) with it's [AppImage plugin](https://github.com/linuxdeploy/linuxdeploy-plugin-appimage), where such implicitness doesn't exist, everything's opt-in.

**Comment by TheAssassin** [comment]
> You can read it out with:
> 
> ```
> objdump -s -j .upd_info *.AppImage
> ```

You can also call the AppImage in question with `--appimage-updateinformation`, if you trust it (note: only works for type 2).

**Comment by technic** [comment]
@TheAssassin cool, that's good to know!

**Comment by technic** [comment]
How to use bintray-zsync?
Should the file uploaded to bintray have the same name for all version, or should I have the same name for AppImage and AppImage.zsync files and override them each time?
Now I have a problem that my files have pattern: "application-$VERSION-x68_64.Appimage", but when I launch appimageupdatetool it looks for application-x68_64.Appimage and fails with 404.

**Comment by probonopd** [comment]
See https://bintray.com/probono/AppImages/Firefox#files for an example.

**Comment by technic** [comment]
@probonopd, thanks for quick reply! As far as I can see each version has its own file in your case.
Then I go to the 2nd part of my question: how to tell appimageupdate to look for the correct AppImage url?
Now it downloads zsync file correctly, but then it asks for wrong url in zsync:
```bash
$ strings e2designer-v0.1-205-g230c48e-x86_64.AppImage.zsync |grep URL
URL: e2designer-x86_64.AppImage
```
You have:
```
$ strings Firefox-63.0.3.glibc2.7-x86_64.AppImage.zsync |grep URL
URL: http://dl.bintray.com/probono/AppImages/Firefox-63.0.3.glibc2.7-x86_64.AppImage
```
I expect the URL to contain version as well. Is it correct? How to set it? Currently I am using
```bash
appimagetool -u "https://api.bintray.com/content/technic/e2designer/e2designer/e2designer-_latestVersion-x86_64.AppImage.zsync" appdir
```
to generate the files. So url for zsync file itself is correct, though inside zsync file I have a link to non-existing file.

**Comment by probonopd** [comment]
`VERSION=$... .(appimagetool-*.AppImage -n --bintray-user $BINTRAY_USER --bintray-repo $BINTRAY_REPO -v ./$APP.AppDir/`

**Comment by technic** [comment]
@probonopd, Thanks for your answers, it works now. But do we benefit from rsync this way? And is it safe to remove old versions?

**Comment by TheAssassin** [comment]
Of course you benefit of the rsync algorithm. Yes, it's safe to remove old versions. Think "we take the local file, check for the bits which have not changed, copy them into a new file, and fill the rest with the data from the latest file on the server". "Latest file" here means "whatever the `.zsync` file your update information points to describes". Some update information types, like bintray or github, provide some magic "latest" aliases which make use of the respective APIs to get the latest `.zsync` file. Otherwise, you usually just create a static (as in: filename is fixed) symlink what you want to be the "latest" `.zsync` file. This can be automated.

**Comment by probonopd** [comment]
Does it work for you @technic, can we close this issue?

**Comment by technic** [comment]
Yes, your command works perfect!

**Comment by TheAssassin** [comment]
Feel free to open new issues or better join our IRC chat (#appimage on Freenode) if you have more questions.
